addons/payment/controllers/post_processing.py

```diff
@@ -47,6 +47,9 @@ def poll_status(self, **_kwargs):
         if not monitored_tx:  # The session might have expired, or the tx has never existed.
             raise Exception('tx_not_found')
 
+        if request.env.user._is_public():
+            monitored_tx = monitored_tx.with_company(monitored_tx.company_id)
+
```

Comment:
Global fixes like that should be avoided, can't you simply make sure the necessary operation is done with the right company at all times, regardless of whether the user is public or not?

It's not clear from your explanation but IIUC, the issue comes from the post processing of transactions in `account_payment` and it either comes from `action_post` or `_create_payment`.  Depending on which one is the root of the issue:
* if it's the `_create_payment` call, you can easily call it with the right company of the transaction.
* if it's the `action_post` call, you'll have to split the recordset of transactions by company, and iterate on it to call the method with the right company.

NB: in 17.3+, `self.invoice_ids` is used while looping on self, it could be changed to `tx.invoice_ids` in your forward-port.